language/move-compiler/src/typing/core.rs

```diff
@@ -512,6 +514,15 @@ fn error_format_impl_(b_: &Type_, subst: &Subst, nested: bool) -> String {
             let inner = format_comma(tys.iter().map(|s| error_format_nested(s, subst)));
             format!("({})", inner)
         }
+        Apply(_, sp!(_, TypeName_::Builtin(sp!(_, BuiltinTypeName_::Fun))), tys) => {
+            assert!(!tys.is_empty(), "ICE invalid function type");
+            let k = tys.len() - 1;
+            format!(
```

Comment:
Stuff like this is I think a perfect reason to make a new `Type` variant :)

Reply:
I tend to disagree. The price you pay is that all kind of algorithms which match over types need to be extended.  After all, logically this is a type constructor; it only happens to have a special user syntax (similar as in Rust).